processing/src/main/java/io/druid/segment/filter/ExpressionFilter.java

```diff
@@ -0,0 +1,120 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.filter;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import io.druid.math.expr.Evals;
+import io.druid.math.expr.Expr;
+import io.druid.math.expr.Parser;
+import io.druid.query.BitmapResultFactory;
+import io.druid.query.expression.ExprUtils;
+import io.druid.query.filter.BitmapIndexSelector;
+import io.druid.query.filter.Filter;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import io.druid.segment.ColumnSelector;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.virtual.ExpressionSelectors;
+
+import java.util.Set;
+
+public class ExpressionFilter implements Filter
+{
+  private final Expr expr;
+  private final Set<String> requiredBindings;
+
+  public ExpressionFilter(final Expr expr)
+  {
+    this.expr = expr;
+    this.requiredBindings = ImmutableSet.copyOf(Parser.findRequiredBindings(expr));
+  }
+
+  @Override
+  public ValueMatcher makeMatcher(final ColumnSelectorFactory factory)
+  {
+    final LongColumnSelector selector = ExpressionSelectors.makeLongColumnSelector(factory, expr, 0L);
+    return new ValueMatcher()
+    {
+      @Override
+      public boolean matches()
+      {
+        return Evals.asBoolean(selector.get());
+      }
+
+      @Override
+      public void inspectRuntimeShape(final RuntimeShapeInspector inspector)
+      {
+        inspector.visit("selector", selector);
+      }
+    };
+  }
+
+  @Override
+  public boolean supportsBitmapIndex(final BitmapIndexSelector selector)
+  {
+    // We can operate on bitmap indexes if this is a constant expression, or if it reads one column that has
+    // an index and does not have multiple values. The lack of multiple values is important because expression
+    // filters treat multi-value arrays as nulls, which doesn't permit index based filtering.
+    return requiredBindings.isEmpty()
+           || (requiredBindings.size() == 1
+               && selector.getBitmapIndex(Iterables.getOnlyElement(requiredBindings)) != null
+               && !selector.hasMultipleValues(Iterables.getOnlyElement(requiredBindings)));
+  }
+
+  @Override
+  public <T> T getBitmapResult(final BitmapIndexSelector selector, final BitmapResultFactory<T> bitmapResultFactory)
+  {
+    if (requiredBindings.isEmpty()) {
+      // Constant expression.
+      if (expr.eval(ExprUtils.nilBindings()).asBoolean()) {
+        return bitmapResultFactory.wrapAllTrue(Filters.allTrue(selector));
+      } else {
+        return bitmapResultFactory.wrapAllFalse(Filters.allFalse(selector));
+      }
+    } else {
+      // Can assume there's only one binding and it has a bitmap index, otherwise the caller should not have called us.
```

Comment:
Because callers should not call `getBitmapIndex` without checking `supportsBitmapIndex` first. I'll add that to the javadocs for Filter since they're not there right now. All the existing callers do satisfy this.